activerecord/lib/active_record/associations/preloader.rb

```diff
@@ -98,26 +98,33 @@ def preload(records, associations, preload_scope = nil)
       private
 
         # Loads all the given data into +records+ for the +association+.
-        def preloaders_on(association, records, scope)
+        def preloaders_on(association, records, scope, polymorphic_parent = false)
           case association
           when Hash
-            preloaders_for_hash(association, records, scope)
+            preloaders_for_hash(association, records, scope, polymorphic_parent)
           when Symbol
-            preloaders_for_one(association, records, scope)
+            preloaders_for_one(association, records, scope, polymorphic_parent)
           when String
-            preloaders_for_one(association.to_sym, records, scope)
+            preloaders_for_one(association.to_sym, records, scope, polymorphic_parent)
           else
             raise ArgumentError, "#{association.inspect} was not recognized for preload"
           end
         end
 
-        def preloaders_for_hash(association, records, scope)
+        def preloaders_for_hash(association, records, scope, polymorphic_parent)
           association.flat_map { |parent, child|
-            loaders = preloaders_for_one parent, records, scope
+            loaders = preloaders_for_one parent, records, scope, polymorphic_parent
 
             recs = loaders.flat_map(&:preloaded_records).uniq
+            polymorphic_parent = begin
+              records.first.association(parent).options[:polymorphic]
```

Comment:
What do you think about instead of relying on the exception we check if the association exists?

```
reflection = records.first.class._reflect_on_association(parent)
polymorphic_parent = if reflection
  records.first.association(parent).options[:polymorphic]
else
  raise AssociationNotFoundError.new(record.first, parent) unless polymorphic_parent
  false
end
```